spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/LayerNormalization.scala

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl._
+import com.intel.analytics.bigdl.nn.abstractnn.{Activity, TensorModule}
+import com.intel.analytics.bigdl.nn.{Module => _}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect.ClassTag
+
+/**
+ * Applies layer normalization.
+ * @param hiddenSize
+ * @param ev$1
+ * @param ev
+ * @tparam T
+ */
+class LayerNormalization[T: ClassTag](hiddenSize: Int)
+  (implicit ev: TensorNumeric[T]) extends BaseModule[T] {
+  override def buildModel(): Module[T] = {
+    val input = Input()
+    val mean = Mean(-1, squeeze = false).inputs(input)
+    val sub = CSubTableExpand().inputs(input, mean)
+    val square = Square().inputs(sub)
+    val mean2 = Mean(-1, squeeze = false).inputs(square)
+    val add = AddConstant(1e-6).inputs(mean2)
+    val sqrt = Power(-0.5, 1, 0).inputs(add)
+    val mul = CMulTableExpand().inputs(sub, sqrt)
+    val linear = new LayerLinear[T](hiddenSize).inputs(mul)
+    Graph(input, linear)
+  }
+  override def updateOutput(input: Activity): Activity = {
+    output = model.updateOutput(input)
+    output
+  }
+}
+
+/**
+ * Implement x * weight vector + bias vector
+ * @param hiddenSize
+ * @param ev$1
+ * @param ev
+ * @tparam T The numeric type in this module parameters
+ */
+private[nn] class LayerLinear[T: ClassTag](val hiddenSize: Int)
```

Comment:
What's the difference with `Linear` ?

Reply:
This layer implements `input vector * weight vector + bias vector`. It should be an element-wise multiply operation.
If input is a tensor with two, then we should select tensors along its's first dim to do multiply.

So you see it is different with nn linear. Maybe name is not clear. How about `VectorLinear` or any other?

Reply:
`VectorProduct` ?

Reply:
ok, changed